[Blaze] Learn how to blaze sheet UI

selanthiraiyan opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #11566 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
Presents the learn how sheet from the new Blaze intro screen.

## Testing instructions

- Create a JN site
- Create a new product and publish it
- From the Dashboard Blaze view tap "Create Campaign" button
- You should see the new intro view
- Tap "Learn how Blaze works" button
- You will see the new sheet

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->
| Light | Dark |
|--------|--------|
| ![Simulator Screen Recording - iPhone 14 Pro - 2023-12-29 at 18 44 54](https://github.com/woocommerce/woocommerce-ios/assets/524475/e5ba6183-beec-4212-b571-e1a52d655715) | ![Simulator Screen Recording - iPhone 14 Pro - 2023-12-29 at 18 45 09](https://github.com/woocommerce/woocommerce-ios/assets/524475/d85ea084-9774-454e-ab8f-2f36a0ceddd1) | 

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

selanthiraiyan committed:
Add SwiftUI view for Blaze learn how sheet.

selanthiraiyan committed:
Add view model to present learn how sheet.

selanthiraiyan committed:
Present learn how to blaze sheet.

selanthiraiyan committed:
Merge branch 'trunk' into feat/11566-learn-blaze-sheet

selanthiraiyan committed:
Update blaze campaign illustration image.

selanthiraiyan committed:
Stop manually adding "$5" text to the illustration image.

selanthiraiyan committed:
Update copy and rearrange features.

hafizrahman reviewed (commented):
There is a flash of scrollbar showing to the right when dragging the sheet up. Is it possible to have it removed? It's maybe similar to the issue here:

https://github.com/woocommerce/woocommerce-ios/pull/11507#issuecomment-1865858870

hafizrahman commented on WooCommerce/Classes/ViewRelated/Blaze/BlazeIntro/BlazeCreateCampaignIntroViewModel.swift:
```diff
@@ -0,0 +1,17 @@
+final class BlazeCreateCampaignIntroViewModel: ObservableObject {
+    private let analytics: Analytics
+
+    @Published var showLearHowSheet: Bool = false
```
nit: typo: "showLearnHowSheet"

selanthiraiyan replied:
Nice catch. Fixed in https://github.com/woocommerce/woocommerce-ios/pull/11583/commits/4d0269eb6eb0a227979258b57a712dba3e85f739

itsmeichigo commented:
👋  I believe the close button should be consistent between the screens:

<img width="272" alt="Screenshot 2024-01-02 at 09 19 42" src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/86162cf5-fcfd-437f-b91c-7eb5adbb3edc">

selanthiraiyan committed:
[Blaze] Intro view image and copy update (#11584)

selanthiraiyan committed:
Merge branch 'trunk' into feat/11566-learn-blaze-sheet

selanthiraiyan committed:
Fix typo.

selanthiraiyan committed:
Remove circle background from close button.

selanthiraiyan committed:
Scale circle background.

selanthiraiyan commented:
> 👋 I believe the close button should be consistent between the screens:

It looks like the design has been updated recently. I have updated the close button in https://github.com/woocommerce/woocommerce-ios/pull/11583/commits/787fd52d7e099ea9c1611533d62d913726476079

selanthiraiyan commented:
> There is a flash of scrollbar showing to the right when dragging the sheet up. Is it possible to have it removed? It's maybe similar to the issue here:

That is correct. This behaviour is similar to the https://github.com/woocommerce/woocommerce-ios/pull/11507#issuecomment-1865858870. The scroll indicator shows up when we drag the sheet and scroll view together. I don't think this needs to be fixed.

selanthiraiyan commented:
Thanks for the review, @hafizrahman and @itsmeichigo! I have addressed the comments and the PR is ready for another look. 🙏

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Blaze/BlazeIntro/BlazeLearnHowView.swift:
```diff
@@ -0,0 +1,242 @@
+
+private extension BlazeLearnHowView {
+    var titleBlock: some View {
+        ZStack(alignment: .center) {
```
I don't think we need to use `ZStack` in this case. How about a simple `HStack` for both the button and the title?

```Swift
HStack {
    Button(action: {}, label: { ... })
    Spacer()
    Text(String(format: Localization.title, "Blaze"))
         .fontWeight(.semibold)
          .headlineStyle()
    Spacer()
}
.padding(.horizontal, Layout.TitleBlock.closeButtonHorizontalPadding)
.padding(.top, Layout.TitleBlock.topPadding)

```

selanthiraiyan replied:
I used a `ZStack` because this setup allows us to align the title text at the center. If we use a `HStack`, the title text won't be perfectly centered. What do you think?

itsmeichigo replied:
You're right, I tried the solution and found that the title is not perfectly centered. What I don't like about ZStack or `overlay` is that the views can overlap each other (e.g when the text is long/large enough):
<img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/f6715f6e-d380-48cc-95f7-81c3edf04ac4" width=320 />

I was thinking maybe we could use a navigation bar and [hide the background](https://developer.apple.com/documentation/swiftui/view/toolbarbackground(_:for:)-7lv0f), but the API is not available until iOS 16. 

I don't have other suggestions for this.

selanthiraiyan replied:
> What I don't like about ZStack or overlay is that the views can overlap each other (e.g when the text is long/large enough):

Good point about the overlapping. I replaced the `ZStack` with `HStack` as per your suggestion. https://github.com/woocommerce/woocommerce-ios/pull/11583/commits/80b8b8fa51bb91430615e745418c7dd0318ba75e

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Blaze/BlazeIntro/BlazeLearnHowView.swift:
```diff
@@ -0,0 +1,242 @@
+
+    var stepBlock: some View {
+        ScrollView {
+            VStack(alignment: .leading, spacing: 0) {
```
This extra VStack is redundant, we should be able to apply the corner radius to the inner one.

selanthiraiyan replied:
Good point. Removed in https://github.com/woocommerce/woocommerce-ios/pull/11583/commits/79e4dc8a28292333e53a9df5631de1adcaa9d4a5

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Blaze/BlazeIntro/BlazeLearnHowView.swift:
```diff
@@ -0,0 +1,242 @@
+            comment: "Title for the Blaze Learn How screen. %1$@ is replaced with string Blaze. Reads like: Learn how Blaze works"
+        )
+        enum ChooseProduct {
+            static let number = NSLocalizedString(
+                "blazeLearnHowView.chooseProduct.stepNumber",
+                value: "1",
+                comment: "Step number for Choose a product step"
+            )
```
Please remove the localization for the numbers as they should stay fixed.

selanthiraiyan replied:
Done in https://github.com/woocommerce/woocommerce-ios/pull/11583/commits/5697ce1a49863666dbdac981540c3541b9e3b1ae

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Blaze/BlazeIntro/BlazeLearnHowView.swift:
```diff
@@ -0,0 +1,242 @@
+
+private extension BlazeLearnHowView {
+    struct Step {
+        let number: String
```
Should number be an Int instead of String?

selanthiraiyan replied:
Updated in https://github.com/woocommerce/woocommerce-ios/pull/11583/commits/5697ce1a49863666dbdac981540c3541b9e3b1ae

itsmeichigo reviewed (changes requested):
The UI matches the design now ✅ I left some comments regarding the implementation below.

selanthiraiyan committed:
Remove redundant VStack.

selanthiraiyan committed:
Stop localizing step numbers.

selanthiraiyan commented:
Thanks for the suggestions, @itsmeichigo! I pushed the necessary updates. I left a  https://github.com/woocommerce/woocommerce-ios/pull/11583#discussion_r1439156127 about the usage of ZStack for the title block. Let me know what you think.

selanthiraiyan committed:
Merge branch 'trunk' into feat/11566-learn-blaze-sheet

itsmeichigo reviewed (approved):
Thanks for the updates! I added a comment regarding the navigation bar - I don't have the perfect answer for the solution, so feel free to decide what's best for the case.

selanthiraiyan committed:
Use HStack to avoid overlapping.

selanthiraiyan commented:
> I added a comment regarding the navigation bar - I don't have the perfect answer for the solution, so feel free to decide what's best for the case.

As you suggested, I replaced the ZStack with HStack to avoid the overlapping issue. Enabling auto merge now. Please feel free to comment if any. Thanks again for the review.
